feat(broker): add redis Pub/Sub broker

brian-gavin opened the pull request:
Adds the redis pub/sub broker to the server. Worth some more looking
into, there might be some rate limiting etc. that is done. This just
adds simple support for it using a single channel that is subscribed and
published to.

brian-gavin committed:
feat(broker): add redis Pub/Sub broker

Adds the redis pub/sub broker to the server. Worth some more looking
into, there might be some rate limiting etc. that is done. This just
adds simple support for it using a single channel that is subscribed and
published to.

rmushkot commented on server/daemon/broker/redis/redis.go:
```diff
@@ -0,0 +1,93 @@
+
+// Subscribe prepares the peer to consume messages.
+func (p *PubSubPeer) Subscribe() error {
+	p.pubsub = p.conn.Subscribe(context.Background(), channelKey)
```
Is this synchronous by default? Could be the cause of poor results.

brian-gavin replied:
I'm not sure. I'll have to look into it - though it might be. In terms of perf, wouldn't subscribe only be called once per consumer?

rmushkot commented on server/daemon/broker/redis/redis.go:
```diff
@@ -0,0 +1,93 @@
+		for {
+			select {
+			case msg := <-p.send:
+				err := p.conn.Publish(context.Background(), channelKey, msg).Err()
```
Is this synchronous by default? Could be the cause of poor results.

brian-gavin replied:
Again I'll have to look, though this could be thrown into a goroutine.

brian-gavin replied:
Putting this into a goroutine made the producer throughput increase a couple 10x

brian-gavin committed:
throw publish routine in a goroutine for ez gains

brian-gavin committed:
fix(redis): docker properly kills the container

brian-gavin committed:
limit number of goroutines that publish for redis

Limited by the pool size of the redis client, this seems like a decent
size to limit on as, accoring to the Redis docs, there are 10
connections per CPU. this seems stable (my machine can handle the 1 mill
messages case) and performant (still an increase over single publisher
goroutine)
